src/en/news/blog/2022/mclock-vs-wpq-testing-with-background-ops-part2/index.md

```diff
@@ -0,0 +1,266 @@
+
+![](images/image15_IOPS_with_modified_cost.png)
+
+## Client Ops with Background Scrub & Recovery
+
+### Test Steps
+
+1. Bring up the Ceph cluster with 4 OSDs
+2. Create a client RBD pool with replication factor of 3 and prefill some objects into it.
+3. Create a RBD recovery pool with replication factor 3.
+4. Mark an OSD down and out. After the cluster handles the OSD down event, 50K 4MiB objects are pre-filled into the recovery pool.
+5. Create a RADOS scrub pool with replication factor 3
+6. Temporarily disable recovery and backfill ops.
+7. Bring up the downed OSD from step 4
+8. Prefill the scrub pool with 100000 objects of 4 MiB size
+9. Initiate deep scrub on the scrub pool and enable recovery and backfill at the same time
+10. Initiate I/O on the client pool for 9 minutes using fio
+11. While the client I/O and deep scrub are running, the test collects statistics related to client throughput and latency. The test also collects scrub statistics such as number of objects scrubbed and scrub duration and recovery statistics.
+
+### Test Results
+
+#### Client Throughput Comparison
+
+![](images/image16_IOPS_mClock_vs_WPQ_recovery_test.png)
+
+The graph above displays the drop in client throughput with recovery and scrub happening in parallel. All the profiles have a similar drop in throughput and the difference between them is not significant. To get a better understanding, we will compare the client latencies in the next sections.
+
+#### Client Latency Comparison
+
+The graph below describes how average client latency varies for the various profiles over the duration of the test.
+
+![](images/image17_avg_clat_while_scrubbing_and_recovery.png)
+
+There is a jump in client latency for the mClock profiles in the beginning of the test. This can be attributed to the beginning of recovery. As we will see in the next section, recovery rate is much higher for mClock profiles as compared to WPQ.
+Over time this spike flattens out and we can see client latency returning to a normal trend and at times even lower than WPQ.
+
+#### Scrub and Recovery Comparison
+
+The graphs below show the scrub and recovery rate over the duration of the test. In the scrub rate comparison, mClock profiles take some time to start scrub whereas with WPQ we can see that scrub starts almost immediately.
+
+![](images/image18_scrub_rate_during_recovery_WPQ_vs_mClock.png)
+
+Almost the opposite trend can be observed in the graph comparing recovery rates. The number of objects recovered starts decreasing rapidly for all the mClock profiles. WPQ seems to have a slow start and eventually takes much longer to complete recovery than the mClock profiles.
+
+![](images/image19_recovery_rate_during_scrub_WPQ_vs_mClock.png)
+
+## Conclusion
+
+To summarize, after taking a look at how the default mClock profiles handle background operations such as scrubbing and recovery, we can say that mClock profiles attempt to finish recovery faster than a scrub.
```

Comment:
Is this a way of saying that mClock profiles attempt to prioritize recovery operations over scrub? I am just a little unclear about what this sentence means.

Reply:
Thanks! I'll work on rephrasing it

Reply:
Instead of adding more details to the conclusion, I have added some explanation in the section before. Let me know if that looks better now.

Reply:
Looks great!